Foundation/URLCache.swift

```diff
@@ -173,13 +217,102 @@ open class URLCache : NSObject {
     */
     open class var shared: URLCache {
         get {
-            NSUnimplemented()
+            return sharedLock.performLocked {
+                return _shared
+            }
         }
         set {
-            NSUnimplemented()
+            sharedLock.performLocked {
+                _shared = newValue
+            }
         }
     }
-
+    
+    private let cacheDirectory: URL?
+    
+    private struct CacheEntry: Hashable {
+        var identifier: String
+        var cachedURLResponse: CachedURLResponse
+        var date: Date
+        var cost: Int
+        
+        init(identifier: String, cachedURLResponse: CachedURLResponse, serializedVersion: Data? = nil) {
+            self.identifier = identifier
+            self.cachedURLResponse = cachedURLResponse
+            self.date = Date()
+            // Estimate cost if we haven't already had to serialize this.
+            self.cost = serializedVersion?.count ?? (cachedURLResponse.data.count + 500 * (cachedURLResponse.userInfo?.count ?? 0))
+        }
+        
+        func hash(into hasher: inout Hasher) {
+            hasher.combine(identifier)
+        }
+        
+        static func ==(_ lhs: CacheEntry, _ rhs: CacheEntry) -> Bool {
+            return lhs.identifier == rhs.identifier
+        }
+    }
+    
+    private let inMemoryCacheLock = NSLock()
+    private var inMemoryCacheOrder: [String] = []
+    private var inMemoryCacheContents: [String: CacheEntry] = [:]
+    
+    func evictFromMemoryCacheAssumingLockHeld(maximumSize: Int) {
+        let sizes: [Int] = inMemoryCacheOrder.map {
+            inMemoryCacheContents[$0]!.cost
+        }
+        
+        var totalSize = sizes.reduce(0, +)
+        
+        guard totalSize > maximumSize else { return }
+        
+        var identifiersToRemove: Set<String> = []
+        for (index, identifier) in inMemoryCacheOrder.enumerated() {
+            identifiersToRemove.insert(identifier)
+            totalSize -= sizes[index]
+            if totalSize < maximumSize {
+                break
+            }
+        }
+        
+        for identifier in identifiersToRemove {
+            inMemoryCacheContents.removeValue(forKey: identifier)
+        }
+        inMemoryCacheOrder.removeAll(where: { identifiersToRemove.contains($0) })
```

Comment:
@drodriguez I completely forgot about `dropFirst`, which obviously is always correct here. I'll adopt your suggestion.